onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -177,10 +177,12 @@ static Status GetCapabilityForEP(const GetCapabilityForEPParams& params) {
   }
 
 #if !defined(ORT_MINIMAL_BUILD) || defined(ORT_EXTENDED_MINIMAL_BUILD)
-  // Run layout transformer only for EPs other than CPU EP and provided the preferred layout is NHWC
-  // CPU EP layout transformation happens later when level 3 transformers are run.
-  if (params.mode != GraphPartitioner::Mode::kAssignOnly &&
-      current_ep.GetPreferredLayout() == DataLayout::NHWC) {
+  // Run layout transformation for all EPs.
+  // For an EP that wants NHWC this will wrap layout sensitive nodes with Transpose nodes first.
+  // In both NCHW and NHWC EPs the EP specific transpose optimization is run last to optimize
+  // transposes for nodes assigned to the EP or unassigned nodes. This allows things like the
+  // EP aware Resize handling to be run.
+  if (params.mode != GraphPartitioner::Mode::kAssignOnly && params.transform_layout.get()) {
```

Comment:
I have two questions here.

1. Shouldn't layout transformation (more specifically transpose optimization) be run for NCHW EPs (e.g., cuda) _only_ if L1 optimizations are enabled?
2. In order to prevent this from running for all EPs, can we pass in a `has_layout_sensitive_ops` function with the params argument? The `has_layout_sensitive_ops` could be a lambda that wraps the existing `GetEPLayoutSensitiveOps` function.

```C++
// In inference_session.cc
auto has_layout_sensitive_ops_fn = [](const std::string& ep) -> bool {
  return !layout_transformation::GetEPLayoutSensitiveOps(ep).empty();
};

partitioner.Partition(graph, ..., transform_layout_fn, mode, debug_graph_fn, has_layout_sensitive_ops_fn);

// In graph_partitioner.cc::GetCapabilityForEP()
const bool ep_prefers_nhwc = current_ep.GetPreferredLayout() == DataLayout::NHWC;
const bool ep_needs_transpose_opt = params.has_layout_sensitive_ops.get() && params.has_layout_sensitive_ops(current_ep.Type());

if (params_mode != kAssignOnly && params.transform_layout.get() && (ep_prefers_nhwc || ep_needs_transpose_opt)) {
  // Run layout transformation
}

Reply:
Or instead of passing in a `has_layout_sensitive_ops` function, we can directly pass in a `ep_needs_layout_transformation` function that bakes in all checks (preferred layout is nhwc or has layout sensitive ops with L1 opts enabled).

Reply:
Just realized that `GetEpLayoutSensitiveOps` also returns FuseConv and others, so that wouldn't exactly work...

Reply:
1. Yeah I guess. We wouldn't have run the transpose optimizer prior to partitioning so it would be more consistent to not suddenly run it here. Maybe there should also be a check saying you cannot use an EP that requests NHWC layout if level 1 optimizations are disabled. 

2. I would have expected any meaningful EP to handle layout sensitive ops so not clear how much value an extra function adds. Is it a performance concern? If the EP didn't have layout sensitive ops we wouldn't change the model, and the L1 optimizations also wouldn't run (they only do if the graph is modified).

Reply:
> I would have expected any meaningful EP to handle layout sensitive ops so not clear how much value an extra function adds. Is it a performance concern? 

I primarily suggested this extra function because your original comment on this code noted that "this feels a little too heavy as it only needs to be called for NCHW EPs with layout sensitive Resize currently." I was trying to suggest something that would avoid having to run an unnecessary layout transformation on EPs that don't need it. However, I'm not sure if this would even provide a meaningful performance improvement, just throwing it out there.